File: Unique Binary Search Trees II.cpp
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<TreeNode *> generateTrees(int n) {
        // Start typing your C/C++ solution below
        // DO NOT write int main() function
        return generateTrees(1, n);
    }
    
    vector<TreeNode *> generateTrees(int left, int right)
    {
        vector<TreeNode *>result;
        if(left > right)
        {
            TreeNode *node = NULL;
            result.push_back(node);
        }
        else
        {
            TreeNode *root = NULL;
            for(int i = left; i <= right; i++)
            {
                
                vector<TreeNode *> leftList = generateTrees(left, i-1);
                vector<TreeNode *> rightList = generateTrees(i+1, right);
                for(TreeNode *&left:leftList)
                {
                    for(TreeNode *&right:rightList)
                    {
                        root = new TreeNode(i);
                        root->left = left;
                        root->right = right;
                        result.push_back(root);
                    }
                }
                
            }
        }
        
        return result;
    }
};

File: Unique Binary Search Trees II.java
/**
 * Definition for binary tree
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; left = null; right = null; }
 * }
 */
public class Solution {
    public ArrayList<TreeNode> generateTrees(int n) {
        // Start typing your Java solution below
        // DO NOT write main() function
        return generateTrees(1, n);
    }
    
    public ArrayList<TreeNode> generateTrees(int start, int end)
    {
        ArrayList<TreeNode> ret = new ArrayList<TreeNode>();  
        if(start > end)
        {
            ret.add(null);
            return ret;
        }
        
        for(int i = start; i < end + 1; i++)
        {
            ArrayList<TreeNode> leftBranches = generateTrees(start, i-1);
            ArrayList<TreeNode> rightBranches = generateTrees(i+1, end);
            for(TreeNode left:leftBranches)
            {
                for(TreeNode right:rightBranches)
                {
                    TreeNode root = new TreeNode(i);
                    root.left = left;
                    root.right = right;
                    ret.add(root);
                }
            }
            
        }
        
        return ret;
    }
    
}